src/sinks/postgres/service.rs

```diff
@@ -0,0 +1,140 @@
+
+    fn is_retriable_error(&self, _error: &Self::Error) -> bool {
+        // TODO: Implement this
+        false
+    }
+}
+
+#[derive(Clone)]
+pub struct PostgresService {
+    connection_pool: Pool<Postgres>,
+    table: String,
+    endpoint: String,
+}
+
+impl PostgresService {
+    pub const fn new(connection_pool: Pool<Postgres>, table: String, endpoint: String) -> Self {
+        Self {
+            connection_pool,
+            table,
+            endpoint,
+        }
+    }
+}
+
+// TODO: do we need this clone?
+#[derive(Clone)]
+pub struct PostgresRequest {
+    pub events: Vec<LogEvent>,
+    pub finalizers: EventFinalizers,
+    pub metadata: RequestMetadata,
+}
+
+impl From<Vec<LogEvent>> for PostgresRequest {
+    fn from(mut events: Vec<LogEvent>) -> Self {
+        let finalizers = events.take_finalizers();
+        let metadata_builder = RequestMetadataBuilder::from_events(&events);
+        let events_size = NonZeroUsize::new(events.estimated_json_encoded_size_of().get())
+            .expect("payload should never be zero length");
+        // TODO: is this metadata creation correct?
+        let metadata = metadata_builder.with_request_size(events_size);
+        PostgresRequest {
+            events,
+            finalizers,
+            metadata,
+        }
+    }
+}
+
+impl Finalizable for PostgresRequest {
+    fn take_finalizers(&mut self) -> EventFinalizers {
+        self.finalizers.take_finalizers()
+    }
+}
+
+impl MetaDescriptive for PostgresRequest {
+    fn get_metadata(&self) -> &RequestMetadata {
+        &self.metadata
+    }
+
+    fn metadata_mut(&mut self) -> &mut RequestMetadata {
+        &mut self.metadata
+    }
+}
+
+pub struct PostgresResponse {
+    metadata: RequestMetadata,
+}
+
+impl DriverResponse for PostgresResponse {
+    fn event_status(&self) -> EventStatus {
+        EventStatus::Delivered
+    }
+
+    fn events_sent(&self) -> &GroupedCountByteSize {
+        // TODO: Is this correct?
+        self.metadata.events_estimated_json_encoded_byte_size()
+    }
+
+    fn bytes_sent(&self) -> Option<usize> {
+        Some(self.metadata.request_encoded_size())
+    }
+}
+
+impl Service<PostgresRequest> for PostgresService {
+    type Response = PostgresResponse;
+    type Error = PostgresError;
+    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
+
+    fn poll_ready(&mut self, _cx: &mut Context) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, request: PostgresRequest) -> Self::Future {
+        let service = self.clone();
+        let future = async move {
+            let table = service.table;
+            let metadata = request.metadata;
+            // TODO: If a single item of the batch fails, the whole batch will fail its insert.
+            // Is this intended behaviour?
+            sqlx::query(&format!(
+                "INSERT INTO {table} SELECT * FROM jsonb_populate_recordset(NULL::{table}, $1)"
```

Comment:
the table configuration can be a victim of sql injection, but in my opinion, we shouldn't avoid that kind of attacks at this level and the user should be responsible of ensuring that there is not sql injection in the config... The `databend` sink works like this

Reply:
Hm, I suppose `sqlx` does not support parameterized table names? Does the query builder help here? If none of the above works, then we can leave as is.

Reply:
I don't think that could help in this case. See [this statement](https://docs.rs/sqlx/latest/sqlx/struct.QueryBuilder.html#warning-beware-sql-injection-vulnerabilities-and-untrusted-input) about `sqlx`'s query builder. 

And we cannot use a variable bind (`$` syntax) in postgres for table names, as the prepared statements are bounded to a query plan and it cannot change if the target table changes.

I think this is the better way to do it... `sqlx` does not check for sql injection